### DISCUSSION TITLE
How to use this with Mantine?

### DISCUSSION THREAD

**Discussion by VigneshVaidyanathan** [airjp73/rvf#121] [Q&A] [answered]
I am trying to use this library with Mantine and do I have to create a custom component for every form element I use like input, textarea, select etc? to access the input props for each component?

There is no direct example to use Mantine or external UI Library along with Remix Validated Forms. Please help.

**Comment by airjp73** [comment]
Good callout about this not really being documented. That's something that could definitely be worth adding.

In most cases, I think you do most likely want to create a custom component for each Mantine form component you want to use in your forms.

---

Other libraries like react-hook-form provide ways to wire up your forms entirely at the top level without custom components, but I find that approach cumbersome at scale when you have 100+ forms in an app. So this library prioritizes conciseness at the form level at the expense of some extra wiring for each input type.

Now, if you only use one date picker in your app (for example) and really don't want to create a custom component, it's definitely possible to skip that by doing something like this.

```tsx
const { error: myDateError, getInputProps: getDateInputProps } = useField("myDate", { formId: 'my-form' });

return (
  <ValidatedForm validator={validator} id="my-form">
    <MantineDatePicker {...getDateInputProps({ error: myDateError })} />
    <MySubmitButton />
  </ValidatedForm>
)
```

---

Also, depending on your requirements, the form still _works_ even without wiring it up with `useField`. The validation won't be as good as it would otherwise be, though.

The example below will still:
* Validate on form submit
* Submit all the data the same as before

But it won't:
* Validate as the user types
* Clear validation errors until the user attempts to submit again

```tsx
return (
  <ValidatedForm validator={validator}>
    <MantineDatePicker name="myDate" />
    <MySubmitButton />
  </ValidatedForm>
)
```

**Comment by VigneshVaidyanathan** [comment, marked as answer]
I agree when you say it is easier and have more control when we do this way. 

I see that creating a wrapper around the provided Mantine component is not difficult but unwanted. Although it serves the purpose, we are able to get everything wired up properly, it would be better if there is some other workaround that is similar to this [remix-forms](https://remix-forms.seasoned.cc/examples/forms/custom-input) where we get the useField for the entire form control, so we can define it as we want.

Advantages I see when we have that kind of structure

- Easy to wire up without having to write wrapper for every single component we use (if we need the validation to happen as and when user types)
- We can have the schema types which can enforce strongly typing as did in the remix-forms.

I am a beginner when it comes to Remix and how to get these working but if there is a way to get this forms validated without having to write a wrapper, I would be excited. Please let me know.

**Reply by airjp73** [reply]
Are you specifically referring to the render-props api that remix-forms has? That can be achieved in userland pretty easily.

```tsx
const Field = ({ name, children }) => {
  const fieldInfo = useField(name);
  return children(fieldInfo);
}

// Usage
<Field name="myField">
  {({ error, getInputProps }) => <MantineComponent {...getInputProps({ error })} />}
</Field>
```

**Reply by VigneshVaidyanathan** [reply]
Yup thats what I meant. Sorry I am just getting started with React and not sure if this is supposed to be obvious. 

Could you also help me out with this please, 
I have a date and time picker separately i.e. there is a date picker and a time picker to select date and time. So I want to check if both those are set, which I can easily do, but I want to show only one error message

**Reply by airjp73** [reply]
The answer to that is going to depend a bit on which validation library you're using.

There are definitely multiple ways to do this, but here's one suggestion.

If you're using zod, you might build your schema like this:

```tsx
const validator = withZod(
  z.object({
    someUnrelatedField: z.string(),

    // Nest the dates into an object so we can add a validation to the whole object
    dateFields: z.object({
      date: z.string(),
      time: z.string(),
    }).refine(
      dateFields => !!dateFields.date && !!dateFields.time,
      'Please select a date and a time'
    )
  })
);
```

Then in your form, you could grab the error for the `dateFields` object instead of the individual fields.

```tsx
const MyDateTimeField = ({ name }) => {
  const { error } = useField(name);
  return (
    <fieldset>
      <legend>Choose a date and a time</legend>

      {/* MyDatePicker and MyTimePicker can both be using useField internally as well */}
      <MyDatePicker name={`${name}.date`} />
      <MyTimePicker name={`${name}.time`} />

      {error && <MyError>{error}</MyError>
    </fieldset>
  )
}
```

If you're using Yup, you might be able to achieve a similar result with [the test method](https://github.com/jquense/yup#schematestname-string-message-string--function--any-test-function-schema).

**Reply by VigneshVaidyanathan** [reply]
Wow. Thank you so much for the help. I am learning my stuff here so hitting some blocks.

**Comment by kwiat1990** [comment]
I'm also using it with Mantine and I came up with the idea of using `useFormContext` as follows:

```tsx
import { useFormContext, ValidatedForm } from "remix-validated-form";

const formContext = useFormContext("addComment");

return (
  <ValidatedForm
    validator={validator}
    method="post"
    id="addComment"
    defaultValues={{
      name: session?.name
    }}
  >
    <TextInput
      error={formContext.fieldErrors.name}
      defaultValue={formContext.defaultValues?.name}
      label="Name"
      name="name"
    />
  </ValidatedForm>
)
```

**Reply by VigneshVaidyanathan** [reply]
This is a simpler implementation as well.

**Reply by kwiat1990** [reply]
However it seems like it doesn't have all goodies of `getInputProps()` like validation behavior, i.e. validation on blur or change. It works only after submit. @airjp73 is it intended (validation methods must be called explicit on those events) or is this a bug?

**Reply by kwiat1990** [reply]
@VigneshVaidyanathan I have ended up with something like this, which seems a little bit more concise than the initial version with destructured and renamed properties:

```tsx
const formId = "addComment";
const contentField = useField("content", { formId });
const emailField = useField("email", { formId });
const nameField = useField("name", { formId });

...
<TextInput         
  autoFocus={autoFocus}
  label="Name"
  placeholder="Your name"
  size="md"
  {...nameField.getInputProps({
    error: nameField.error
  })}
/>
          
// or you can pass all props directly to `getInputProps` and those are also fully typed!
<Textarea 
  {...contentField.getInputProps({
    "aria-label": "content",
    autoFocus,
    error: contentField.error,
    placeholder: "Add your comment",
    size: "md",
  })}
/>
...
```

**Reply by airjp73** [reply]
Yeah, if you're not using `getInputProps` then you'll need to handle all the validation manually. What you have here with the multiple `useField` calls looks right to me.

**Reply by kwiat1990** [reply]
I have noticed that when I use this approach it's safer to put all component props inside `getInputProps`. Otherwise it can cause issues with input events. In my case `onChange `and `onFocus` didn't work as expected when they defined before `getInputProps`. They should be placed after the function or directly inside it, so:

```tsx
<Textarea 
  // unsafe
  onFocus: ...
  onChange: ...
  {...contentField.getInputProps({
    // safe
    onFocus: ...
    onChange: ...
  })}
  // safe
  onFocus={...}
  onChange={...}
/>
```

**Reply by VigneshVaidyanathan** [reply]
Yes, any functions if you are defining again should be placed after `getInputProps` but I feel having it as a separate field makes it easier for me to make it look clean. I dont want 10 variables to hold the context, it is easier if I can handle it similar to this as mentioned earlier

```
const Field = ({ name, children }) => {
  const fieldInfo = useField(name);
  return children(fieldInfo);
}

// Usage
<Field name="myField">
  {({ error, getInputProps }) => <MantineComponent {...getInputProps({ error })} />}
</Field>

```

**Reply by kwiat1990** [reply]
Hm, ok, what will be the correct typing for such construct:

```tsx
const Field = ({ name, as }: { name: string; as: React.ElementType }) => {
  const { error, getInputProps } = useField(name);
  const Component = as;
  return <Component {...getInputProps({ error })} />;
};

// usage
// TS error: Property 'autosize' does not exist on type 'IntrinsicAttributes & { name: string; as: ElementType<any>;
<Field name="content" as={Textarea} autosize />
```

**Reply by airjp73** [reply]
You're right that the intent of prop getters is to pass everything through the function. Here's an article on the pattern: https://kentcdodds.com/blog/how-to-give-rendering-control-to-users-with-prop-getters

I also want to clarify that the only safe way to handle those call callback props is to pass them to `getInputProps`.

```tsx
<Textarea 
  // unsafe
  onFocus: ...
  onChange: ...

  {...contentField.getInputProps({
    // safe
      onFocus: ...
      onChange: ...
  })}

  // These might feel safe because your callbacks still run,
  // But now your validation behavior is broken
  onFocus={...}
  onChange={...}
/>
```

**Reply by VigneshVaidyanathan** [reply]
Yes. When i use this, i run the validation of that field in my custom on change method i provide.

**Reply by kwiat1990** [reply]
Now it got really interesting ;) I have `RadioGroup` with a few `Radio` component and I really don't know how it should be composed with the `Remix Validated Forms` as `name` and `error` should be set directly on `RadioGroup` and not on radio buttons. In this case something like this won't work:

```tsx
 <RadioGroup name="reasons" defaultValue={reasons[0].key}>
  {reasons.map(({ key, value }) => (
    <Radio
      {...reasonField.getInputProps({
        key: key,
        label: value,
        value: key,
      })}
    />
  ))}
</RadioGroup>
```

I also think the way I set `useField` won't either work.

**Reply by VigneshVaidyanathan** [reply]
Below is how I have used it. I dont remember why I had used `validate` function in the `onChange` implementation but I think it should work as expected. I didnt use the `getInputProps` but kinda used the other `validate` function that validates just that form element.  Let me know if this works for you.


``` typescript
<FormField name="submissionType">
      {({
        error,
        getInputProps,
        validate,
      }: {
        error: string;
        getInputProps: GetInputProps;
        validate: any;
      }) => (
        <RadioGroup
          required
          size="sm"
          label="Entry submission type"
          description="Choose what kind of submissions do you want to accept for the contest."
          className="mt-3 mb-1"
          classNames={{
            error: "text-xs",
            label: "text-sm ",
          }}
          name="submissionType"
          onChange={() => {
            validate();
          }}
          error={error}
          defaultValue={contestDetails?.submissionType}
        >
          <Radio value="picture" label="Photo" />
          <Radio value="video" label="Video" />
          <Radio value="document" label="Document" />
        </RadioGroup>
      )}
    </FormField>
```

**Reply by kwiat1990** [reply]
@VigneshVaidyanathan thanks for the tipp to call `validate()` on `onChange`. That was the last missing part to get it working. I went with my default approach as I can't fully type that `FormField` wrapper component. So my code looks like this:
```tsx
const reasonField = useField("reason", { formId: id });

<RadioGroup
  label="Reason"
  name="reason"
  error={reasonField.error}
  onChange={() => reasonField.validate()}
>
  {reasons.map(({ key, value }) => (
    <Radio
      {...reasonField.getInputProps({
        key,
        label: value,
        size: "md",
        value: key,
      })}
    />
  ))}
</RadioGroup>
```

I was really surprised that a single `useField` did work for multiple radios. I thought I would need to call `useField` directly in the loop like this:

```tsx
 <Radio
  {...useField("reason", { formId: id }).getInputProps({
    key,
    label: value,
    size: "md",
    value: key,
  })}
/>
```

**Reply by VigneshVaidyanathan** [reply]
I am not sure how this might have worked as the getInputProps sets the value, error and handles onChange validation. So technically, in your radio group you arent setting the currently selected value (incase of setting it from the database), in which case you need to send `defaultValue={value}` and this makes it almost that you can forego the `getInputProps` that you have used inside for each Radio and just add simple radio buttons.

The on change event will be fired for the Radio group by Mantine and there you are already handling the validation and that will inturn set the error properties if any. 

Havent tried it in code though so please correct me if I am wrong.

**Reply by FlurryNight** [reply]
Hey, looking for the best way and correct way to do it, can this be documented and an example be provided?

For ex this component: https://mantine.dev/core/password-input/#strength-meter-example

I know i need to create an Custom component using this component inside it, but can I use what the component offers with this package.

This is a very good library , i need to use it and not being able to fully integrate this is delaying my projects

Thank you for your work @airjp73

**Reply by VigneshVaidyanathan** [reply]
Can you explain/ show code of what you have right now? 

As far as I can see, the validation is done outside of Zod Schema which this package uses. So technically, you can have a zod schema that just says `password: zod.string()` and the validation is taken care by the custom implementation.

To get the proper name and other things wired up, you can try to use this implementation 
https://github.com/airjp73/remix-validated-form/discussions/121#discussioncomment-2873082

I havent tried it thought, I can check this and respond later today, if it doesnt work for you,

**Reply by FlurryNight** [reply]
Hi, Thanks for your response , I don't have code right now since i was not able to make it work.

I' will try the implementation u suggested, and report back if smth doesn't work,

Thank you

**Comment by kwiat1990** [comment]
Hi,

I have worked mainly with Vue and I was really surprised that some things in React doesn’t yet exist or are pretty new. Same goes for nested layouts and React Router is similar to Vue Router, which I find really good. 

The thing with using `Box` is that without Tailwind it’s cumbersome to style naked divs. But with it sure, I would also prefer native div with some classes attached to it. 

I‘m working on a blog for my wife with meilisearch and comment system. My backend is Strapi CMS with Postgres and for client I use Remix. All pf this I host on my VPN using Caprover, which is self-hosted PaaS lile heroku and make setup really simple. 

Mateusz

**Comment by intuji-kirankumal** [comment]
@airjp73 I am working on implementing Remix Validated Forms with the Shadcn UI library and Zod Validation Library, particularly focusing on integrating the Shadcn Datepicker component but couldn't find any example. Any guidance or assistance in this process would be greatly appreciated.
